src/auth/src/token_cache.rs

```diff
@@ -74,23 +71,53 @@ impl CachedTokenProvider for TokenCache {
     }
 }
 
+#[async_trait::async_trait]
+impl CachedTokenProvider for TokenCache {
+    async fn token(&self, extensions: Extensions) -> Result<CacheableResource<Token>> {
+        let given_tag = extensions.get::<EntityTag>();
+
+        let (token, tag) = self.get_latest_token_and_entity_tag().await?;
+        if given_tag.is_some_and(|given_tag| tag.eq(given_tag)) {
+            return Ok(CacheableResource::NotModified);
+        } else {
+            return Ok(CacheableResource::New {
+                entity_tag: tag,
+                data: token,
+            });
+        }
+    }
+}
+
 async fn wait_for_next_token(
-    mut rx_token: watch::Receiver<Option<Result<Token>>>,
-) -> Result<Token> {
+    mut rx_token: watch::Receiver<Option<Result<(Token, EntityTag)>>>,
+) -> Result<(Token, EntityTag)> {
     rx_token.changed().await.unwrap();
     let token_result = rx_token.borrow().clone();
 
     token_result.expect("There should always be a token or error in the channel after changed()")
 }
 
-async fn refresh_task<T>(token_provider: T, tx_token: watch::Sender<Option<Result<Token>>>)
-where
+fn get_etag_from_token(token: &str) -> String {
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(token.as_bytes());
+    let hash_bytes = hasher.finalize();
+    format!("\"{}\"", hash_bytes.to_hex())
+}
```

Comment:
If you use my idea with the atomic, this code goes away and you can remove the dependencies.